omnigibson/object_states/particle_modifier.py

```diff
@@ -724,22 +716,20 @@ def _compute_particle_spawn_information(self, system):
         sampling_distance = 2 * system.particle_radius
         extent = np.array(self._projection_mesh_params["extents"])
         h = extent[2]
-        low = np.array([-extent[0] / 2, -extent[1] / 2, 0])
-        high = np.array([extent[0] / 2, extent[1] / 2, h])
-        n_particles_per_axis = ((extent - 2 * system.particle_radius) / sampling_distance).astype(int) + 1
-        arrs = [np.linspace(lo + system.particle_radius, hi - system.particle_radius, n) for lo, hi, n in
-                zip(low, high, n_particles_per_axis)]
+        low = np.array([-extent[0] / 2, -extent[1] / 2, -h])
+        high = np.array([extent[0] / 2, extent[1] / 2, 0])
+        n_particles_per_axis = (extent / sampling_distance).astype(int)
+        assert np.all(n_particles_per_axis), f"link {self.link.name} is too small to sample any particle of radius {system.particle_radius}."
+        # 1e-10 is added because the extent might be an exact multiple of particle radius
+        arrs = [np.arange(lo + system.particle_radius, hi - system.particle_radius + 1e-10, system.particle_radius * 2)
+                for lo, hi, n in zip(low, high, n_particles_per_axis)]
         # Generate 3D-rectangular grid of points, and only keep the ones inside the mesh
         points = np.stack([arr.flatten() for arr in np.meshgrid(*arrs)]).T
-        # Flip the x and z values so that the positions are in the LINK frame, not the MESH frame
-        # This is super ugly, but necessary because omni requires the native visualization XForm to have its
-        # local x-axis pointing in the direction where particles will be spawned!
-        points = points[:, [2, 1, 0]]
         pos, quat = self.link.get_position_orientation()
         points_in_world_frame = get_particle_positions_from_frame(
             pos=pos,
             quat=quat,
-            scale=np.ones(3),
+            scale=self.link.scale,
```

Comment:
we should assert self.obj.scale == self.link.scale

Reply:
and then prob just use self.obj.scale since that's generally what we use elsewhere

Reply:
added to link based state mixin